### ISSUE TITLE
[Feature Request] Add hydra configuration without interpolation

### ISSUE THREAD

**Issue by nirraviv** [allegroai/clearml#327]
Hi,
`OmegaConf` has a powerful tool for interpolating variables in a configuration file(s).
When parsing an `OmegaDict` to YAML or dict (`OmegaConf.to_yaml` or `OmegaConf.to_container`, respectively) you can parse them with or without interpolating the relevant variables with the correct `resolve` argument.

Currently, you are supporting only interpolating variables. The problem may occur when a user runs a task from the server and does not know (or remember) to correctly change all the dependencies. It will be great if you can support both scenarios.

There are two options we were thinking about:
1. Save both configs (in different tabs) - the interpolated for hyperparameters comparison (should be read-only) and the second for running task from the server.
2. Save one of the configurations by request (using a resolving argument).

**Comment by bmartinn** [comment]
Hi @nirraviv 

I think that implementing (1) would be the easiest solution to use, even though might be a bit confusing.
A quick reminder, by default the resolved OmegaConf configuration section is also read-only.

I'm thinking of adding a new `Configuration Object` maybe under `Hydra Configuration` (as the unresolved configuration is actually Hydra's), WDYT?

**Comment by nirraviv** [comment]
Hi @bmartinn ,

In read-only I meant that the resolved configuration should be always read-only and the unresolved should be configurable as today with `_allow_omegaconf_edit_`.

As long as the unresolved configuration will be the editable configuration it can be under `Hydra Configuration`, in my opinion. Will this change the `Hydra Configuration` to look like the current `Configuration Object`? These decisions are more UX than programming.

**Comment by bmartinn** [comment]
> In read-only I meant that the resolved configuration should be always read-only and the unresolved should be configurable as today with _allow_omegaconf_edit_.

Ohh I missed that, really how would you edit the Hydra Conf ? should that be connected with the actual code? Am I missing the use case here ?

> Will this change the Hydra Configuration to look like the current Configuration Object?

Basically, yes. They are both YAML configurations so they are stored as YAML "text file" :)

**Comment by nirraviv** [comment]
Hi @bmartinn , the use case is very simple - when running a task from `clearml-server` don't let the user mistakenly forget to change one of the (originally) interpolated variables and let `OmegaConf` do its magic.

For example if a config file is:
```
var1: 1234
var2: ${var1}
```
I don't want to change both arguments, just `var1`.

**Comment by bmartinn** [comment]
Oh that is an interesting use case!

Now I'm thinking, maybe we should deduce these variables (compare the pre/post resolve dicts) and add them into the Parameters section ? 

Obviously the caveat of the parameter section is that these are flat key/value pairs, but we could do `section/key: value` so we still get the key/value pair in the UI.
wdyt? is it too much automagic?

**Comment by nirraviv** [comment]
Hi @bmartinn ,
> is it too much automatic?

I think so. In my opinion we should keep it as simple as possible to prevent misunderstandings.
Another option is to **not present the interpolated variables in the UI** and just save them unresolved in the background (and clone them when cloning a new task). WDYT?

> BTW: just so I fully understand the use case, you would edit the unresolved section, and expect hydra to reevaluate it, correct?

Indeed, I expect hydra (or OmegaConf) to do their magic. BTW, I think `OmegaDict` is always in the unresolved "version" and interpolate variable in real time.

> Then when you clone a "cloned Task" you are loosing the reference to the original unresolved parameters, no?

This is another reason to save the unresolved config so hydra will be able to interpolate the relevant variables correctly (unless I didn't understand you correctly).

**Comment by bmartinn** [comment]
> Another option is to not present the interpolated variables in the UI and just save them unresolved in the background (and clone them when cloning a new task). WDYT?

So the idea would be that we can edit the unresolved values, correct?
Could you provide a short toy example so we could test (just storing the values is easy, but it might be a bit more challenging to allow to edit the unresolved configuration as opposed the fully resolved one).

> This is another reason to save the unresolved config so hydra will be able to interpolate the relevant variables correctly (unless I didn't understand you correctly).

Hmm, I assume that any unresolved variable (e.g. `var2: ${var1}` , you should never edit, right?
If this is the case I think there is no issue I can see.
(I was thinking what would happen if I manually edit `var2`, then if I have a clone-of-clone then I will never know `var2` was actually a reference to `var1`. Anyhow I think this is out-of-scope for us to solve at the moment 😄  )

**Comment by nirraviv** [comment]
> So the idea would be that we can edit the unresolved values, correct?

Correct. You can add these two dummy variables to your config file in the hydra example for testing. If you need more advanced example I'll send you first thing next week.

> Hmm, I assume that any unresolved variable (e.g. var2: ${var1} , you should never edit, right?

I think we should stick to this assumption for now (otherwise it will be very messy as you stated 😁)

**Comment by nirraviv** [comment]
@bmartinn , any update? Do you need me to send an example or not?

**Comment by bmartinn** [comment]
@nirraviv I think we are good, but if you have a more complicated example that will be appreciated :)
I'm hoping to see an intial commit sometime next week

**Comment by nirraviv** [comment]
@bmartinn any update?

**Comment by bmartinn** [comment]
Hi @nirraviv 
Yes, this is less trivial than expected.
Here is the plan, we are prepping up for a major release this weekend :) 
Once it is out, we will get back to looking into this feature.

**Comment by bmartinn** [comment]
Hi @nirraviv 
Quick update, this feature is already in the latest RC, please let me know what you think.

```
pip install clearml==1.0.3rc0
```

**Comment by nirraviv** [comment]
Hi @bmartinn ,
Looks great. I'm currently working with `hydra-core==1.0.6` and `OmegaConf==2.0.6` but plan to upgrade to `hydra-core==1.1.x` and `OmegaConf==2.1.x` as soon as a stable version will be released. Any idea if this feature will also work for these versions?  check out the release notes for [OmegaConf](https://github.com/omry/omegaconf/releases/tag/v2.1.0.rc1) and [hydra-core](https://github.com/facebookresearch/hydra/releases/tag/v1.1.0.rc1) for more information

**Comment by bmartinn** [comment]
>  Any idea if this feature will also work for these versions? 

Actually it won't, `hydra-core` / `OmegaConf` added the "resolve=False" feature sometime after `1.0.6` (I think).
Anyhow I think 1.1.0 will be released soon 🤞

**Comment by jinmingteo** [comment]
hi @bmartinn, how does one allow _allow_omegaconf_edit_ for Hydra to be set to True?

**Comment by bmartinn** [comment]
@jinmingteo , when you clone the Task in the UI, the Task becomes editable, then go to the configuration Tab, look for the section "Hydra", then double click "_allow_omegaconf_edit_" and change to `True`
